### ISSUE TITLE
Airbrake Not Reporting On Serverless Beta Environment Errors

### ISSUE THREAD

**Issue by robskrob** [airbrake/airbrake-js#560]
I was able to test airbrake locally. Airbrake caught an uncaught exception and reported it -- tagging the environment in which the error occurred as `development`. This is great.

However, when I deploy the same code to a [now lambda](https://zeit.co/now) written in `node`, and I use a different environment variable -- `beta` as the value -- airbrake will not catch this exception and report on it. Has anyone had issues with airbrake reporting on uncaught exceptions locally but not remotely in a deployed environment?

This is my logged out airbrake client in beta if knowing it helps:
```
 2019-07-18T15:27:31.668Z	f2d56d40-71fa-4ef3-97b0-120cf367eda6	airbrake object Client {
                            filters:
                             [ [Function: filter],
                               [Function],
                               [Function: filter],
                               [Function: filter],
                               [Function],
                               [Function: filter] ],
                            offline: false,
                            todo: [],
                            onClose: [],
                            opts:
                             { projectId: #{projectId},
                               projectKey: #{projectKey},
                               environment: 'beta',
                               host: 'https://api.airbrake.io',
                               timeout: 10000,
                               keysBlacklist: [ /password/, /secret/ ] },
                            url: 'https://api.airbrake.io/api/v3/projects/#{projectId}/notices?key=#{someKey}',
                            processor: [Function: processor],
                            requester: [Function: request],
                            historian:
                             Historian {
                               historyMaxLen: 20,
                               notifiers: [ [Circular] ],
                               errors: [],
                               ignoreWindowError: 0,
                               history: [ [Object] ],
                               ignoreNextXHR: 0,
                               consoleError: [Function: prettyConsoleLog],
                               lastState:
                                { type: 'log',
                                  severity: 'log',
                                  arguments: [Array],
                                  date: 2019-07-18T15:26:36.013Z } } }
```

This is the code I have:
```js
var AirbrakeClient = require('airbrake-js');
var airbrakeExpress = require('airbrake-js/dist/instrumentation/express');

var airbrake = new AirbrakeClient({
  projectId: process.env.AIRBRAKE_PROJECT_ID,
  projectKey: process.env.AIRBRAKE_PROJECT_KEY,
  environment:  process.env.ENVIRONMENT
})

const app = express();

app.use(airbrakeExpress.makeMiddleware(airbrake));

//Routes
app.get('/some-error', (req, res) => {
  throw("some errorr!!!!!!!!!!")
})

app.use(airbrakeExpress.makeErrorHandler(airbrake));
```

Whenever I hit the `/some-error` endpoint an exception is thrown but airbrake never catches and reports on the exception in my beta environment as it does locally for me.

Here's my package.json:
```
{
  "name": "nodejs-express",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "airbrake-js": "^1.6.8",
    "axios": "^0.18.0",
    "body-parser": "^1.19.0",
    "cors": "^2.8.5",
    "cross-fetch": "^3.0.4",
    "date-fns": "^1.30.1",
    "dotenv": "^8.0.0",
    "express": "^4.16.4",
    "helmet": "^3.16.0"
  },
  "devDependencies": {
    "chai": "^4.2.0",
    "mocha": "^6.1.4",
    "nock": "^10.0.6",
    "nodemon": "^1.19.0",
    "sanitize-filename": "^1.6.1",
    "sinon": "^7.3.2",
    "testdouble": "^3.11.0"
  },
  "scripts": {
    "start": "node ./index.js",
    "dev": "nodemon index.js",
    "test": "./node_modules/mocha/bin/mocha test/**/*.js"
  }
}
```

**Comment by vmihailenco** [comment]
All looks good and the fact that it works in dev environment does not leave many possibilities. I would check simplest things like if `process.env.AIRBRAKE_PROJECT_ID` has correct value. In your dump it is set to `#{projectId}` but I guess you just replaced it before publishing...

The only other idea I have is that airbrake-js does not work on AWS lambda which is unlikely but possible. I will try to test AWS lambda when we are done with current airbrake-js build refactoring.

**Comment by robskrob** [comment]
@vmihailenco thank you for responding. I will triple check that the AIRBRAKE_PROJECT_ID is accurate, but I'm fairly certain it was what I expected it to be. 

In my logs that I shared in the above issue I set the `projectId`  to `#{projectId}` because I didn't want either my `projectId` or my  `projectKey` to leak on github.

**Comment by bowenli** [comment]
I'm seeing the same thing. On production (rails app), uncaught exceptions are not reported on Chrome/Safari. In Firefox it is working as expected.

**Comment by robskrob** [comment]
@vmihailenco I recreated the same problem on another lambda, `https://testairbrakereporting-jsozb6j0i.now.sh/some-error`. 

Here's the repo: https://github.com/robskrob/airbrake-test-zeit-now

As you can see [here](https://github.com/robskrob/airbrake-test-zeit-now/blob/master/index.js#L33), I'm throwing an error. Locally this is caught in [my project](https://jewell-robertp-gmail-com.airbrake.io/projects/239592) and flagged as the error in development. However when I go to make the same request at `https://testairbrakereporting-jsozb6j0i.now.sh/some-error` -- airbrake does not report on any error.

**Comment by robskrob** [comment]
@vmihailenco it appears other exception reporting services like sentry have had [similar issues](https://github.com/getsentry/sentry-javascript/issues/1449#issuecomment-463682743)

**Comment by zefer** [comment]
Thanks for the extra details, @robskrob. A couple of thoughts that may help debug/resolve this:

1. Does your Lamda's VPC give it access to the public Internet, specifically https://airbrake.io? It's possible that VPC rules are blocking a POST to our servers.
1. ...

> I also try forcibly notifying airbrake and airbrake still does not successfully report on the error.

Perhaps trying [this pattern](https://github.com/airbrake/airbrake-js/blob/master/packages/airbrake-js/CHANGELOG.md#v120) will help shed more light on the problem:

```js
airbrake.notify(err).then(function(notice) {
  if (notice.id) {
    console.log('notice id:', notice.id);
  } else {
    console.log('notify failed:', notice.error);
  }
});
```

**Comment by robskrob** [comment]
@zefer thank you for these suggestions. 

1. I will need to check on this. However, off the top of my head, I do not think the Lambda's VPC is the issue. I have deployed Lambdas that make authenticated HTTP requests to shopify stores publicly available on the internet. Granted, these were GET requests and not POST.
2. I will try that pattern to get more information about the problem

**Comment by zefer** [comment]
@robskrob 👍. [2] might rule-out [1] anyway, so I would start there.

**Comment by robskrob** [comment]
@zefer I may  have time in the near future to look into this again. Can you please clarify this statement:

> [2] might rule-out [1] anyway, so I would start there.

Are you saying that I should look into option 2 I outlined above?
> 2. I will try that pattern to get more information about the problem

**Comment by zefer** [comment]
> Are you saying that I should look into option 2 I outlined above?

@robskrob Yes, that's where I would start, because it's possible there is an error for a failing HTTP request which would explain what's happening. Possibly not, depending on what the problem is, but it should be relatively quick and easy to try!

**Comment by vmihailenco** [comment]
The problem here is that airbrake-js (and JS generally) sends data via HTTP asynchronously but in serverless environment app exits (paused?) immediately after request is served and airbrake-js does not have any chance to finish sending the error. So the fix is to either send errors synchronously or flush pending errors just before the app exits. To send errors synchronously you can use `await`:

```js
await airbrake.notify("hello world")
```

To flush pending errors you can use `airbrake.flush` which is available in @airbrake/node since v1.0.0-beta.5. The only problem is that it is not clear when and where to call it - it looks like https://zeit.co/ does not provide any suitable hook...

Related PR that tries to solve similar problem is https://github.com/zeit/next.js/issues/8685
